### ISSUE TITLE
RedundantModifierCheck: redundant final on top-level method not flagged in compact source files (JEP 512)

### ISSUE THREAD

**Issue by vivek-0509** [checkstyle/checkstyle#20226]
Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/modifier/redundantmodifier.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="RedundantModifier"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputRedundantModifierCompact.java <<'EOF'
public final void myMethod() { }
void main() { }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml InputRedundantModifierCompact.java

Starting audit...
Audit done.
```

Same code inside an ordinary `final` class (the implicit class in JEP 512 is always `final`):

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > OrdinaryRedundantModifier.java <<'EOF'
public final class OrdinaryRedundantModifier {
    public final void myMethod() { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml OrdinaryRedundantModifier.java

Starting audit...
[ERROR] OrdinaryRedundantModifier.java:2:12: Redundant 'final' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

RedundantModifierCheck must flag `final` on a top-level method in a JEP 512 compact source file. JEP 512 specifies that the implicit class is always `final`, so any `final` modifier on its methods is redundant. The check determines whether a method sits inside a final class by walking its parents looking for `CLASS_DEF`, `LITERAL_NEW`, or `ENUM_CONSTANT_DEF`; for a top-level method in a compact source file none of these is found, so `checkFinal` stays false and the redundant modifier is never flagged.

Expected on the compact input: 1 violation `Redundant 'final' modifier.` on `myMethod`, identical to the ordinary class above.

**Comment by stoyanK7** [comment]
@vivek-0509 Do you think we should also put violation on the `public` modifier?

**Comment by romani** [comment]
@vivek-0509 , please educate us on public modifier

**Comment by vivek-0509** [comment]
@romani @stoyanK7 
Yes, I think flagging public is worth it, and it's consistent with what the check already does. The implicit class in a compact source file is final and it lives in the unnamed package, so it can never be referenced by another class. That means public on its methods can't widen access to anything, it's just obsolete.

So this issue will cover both:

final on top-level methods -> redundant (implicit class is final)
public on top-level methods -> redundant (can't widen access past the unnamed final class)

**Comment by stoyanK7** [comment]
I propose that we limit this issue to the `final` modifier and handle the `public` proposal/discussion in a separate issue. Smaller issues/PRs are _much_ easier to review for us.

@vivek-0509 I see reason in what you said. Please feel free to create a separate issue for the `public` modifier.

**Comment by vivek-0509** [comment]
@stoyanK7 issue created https://github.com/checkstyle/checkstyle/issues/20266

**Comment by romani** [comment]
We have options to skip validation by this Check of such classes at all, and create new Check to handle it more correctly and customizable.

Private  and protected modifier also does not make any sense, just noise in code.

To avoid hammering designs of Check that Target classical types, better to create very special Check.

This Check is already overcomplicated , look at https://checkstyle.org/checks/modifier/redundantmodifier.html#jdkVersion

**Comment by surajgojanur** [comment]
I am on it

**Comment by stoyanK7** [comment]
@surajgojanur, I am not sure there is a clear agreement on what to do here. In the meantime, please feel free to choose some other issue. Just so you don't waste effort on something that is not fully agreed on

**Comment by surajgojanur** [comment]
> [@surajgojanur](https://github.com/surajgojanur), I am not sure there is a clear agreement on what to do here. In the meantime, please feel free to choose some other issue. Just so you don't waste effort on something that is not fully agreed on

okay

**Comment by romani** [comment]
@surajgojanur , please send PR to update doc of RedundantModifier to clearly mention that compact sources are skipped from validation, this will be fix for this issue.

@surajgojanur , please create new issue with new Check design `RedundantModifierCompactSource`.

**Comment by surajgojanur** [comment]
> [@surajgojanur](https://github.com/surajgojanur) , please send PR to update doc of RedundantModifier to clearly mention that compact sources are skipped from validation, this will be fix for this issue.
> 
> [@surajgojanur](https://github.com/surajgojanur) , please create new issue with new Check design `RedundantModifierCompactSource`.

Done.

I opened the documentation PR: #20495

I also created the separate design issue for `RedundantModifierCompactSource`: #20496

**Comment by romani** [comment]
doc is updated to mention that skip is intentional.

RedundantModifierCompactSource https://github.com/checkstyle/checkstyle/issues/20496 will do validation for redundant modifiers on compact sources.
